langchain/src/agents/structured_chat/index.ts

```diff
@@ -0,0 +1,124 @@
+import { StructuredTool } from "../../tools/base.js";
+import { Optional } from "../../types/type-utils.js";
+import { Agent, AgentArgs, OutputParserArgs } from "../agent.js";
+import { AgentInput } from "../types.js";
+import { StructuredChatOutputParserWithRetries } from "./outputParser.js";
+import { FORMAT_INSTRUCTIONS, PREFIX, SUFFIX } from "./prompt.js";
+import { zodToJsonSchema } from "zod-to-json-schema";
+import { JsonSchema7ObjectType } from "zod-to-json-schema/src/parsers/object.js";
+
+export interface ChatCreatePromptArgs {
+  /** String to put after the list of tools. */
+  suffix?: string;
+  /** String to put before the list of tools. */
+  prefix?: string;
+  /** List of input variables the final prompt will expect. */
+  inputVariables?: string[];
+}
+
+export type ChatAgentInput = Optional<AgentInput, "outputParser">;
+
+/**
+ * Agent that interoperates with Structured Tool's using a React logic.
+ * @augments Agent
+ */
+export class StructuredChatAgent extends Agent {
+  constructor(input: ChatAgentInput) {
+    const outputParser =
+      input?.outputParser ?? StructuredChatAgent.getDefaultOutputParser();
+    super({ ...input, outputParser });
+  }
+
+  _agentType() {
+    return "structured-chat-zero-shot-react-description" as const;
+  }
+
+  observationPrefix() {
+    return "Observation: ";
+  }
+
+  llmPrefix() {
+    return "Thought:";
+  }
+
+  _stop(): string[] {
+    return ["Observation:"];
+  }
+
+  static validateTools(tools: StructuredTool[]) {
+    const invalidTool = tools.find((tool) => !tool.description);
+    if (invalidTool) {
+      const msg =
+        `Got a tool ${invalidTool.name} without a description.` +
+        ` This agent requires descriptions for all tools.`;
+      throw new Error(msg);
+    }
+  }
+
+  static getDefaultOutputParser(_fields?: OutputParserArgs, llm?: BaseLanguageModel) {
+    return StructuredChatOutputParserWithRetries.fromLLM(llm);
+  }
+
+  async constructScratchPad(steps: AgentStep[]): Promise<string> {
+    const agentScratchpad = await super.constructScratchPad(steps);
+    if (agentScratchpad) {
+      return `This was your previous work (but I haven't seen any of it! I only see what you return as final answer):\n${agentScratchpad}`;
+    }
+    return agentScratchpad;
+  }
+
+  /**
+   * Create prompt in the style of the zero shot agent.
+   *
+   * @param tools - List of tools the agent will have access to, used to format the prompt.
+   * @param args - Arguments to create the prompt with.
+   * @param args.suffix - String to put after the list of tools.
+   * @param args.prefix - String to put before the list of tools.
+   */
+  static createPrompt(tools: StructuredTool[], args?: ChatCreatePromptArgs) {
+    const { prefix = PREFIX, suffix = SUFFIX } = args ?? {};
+    const toolStrings = tools
+      .map((tool) => `${tool.name}: ${tool.description}, args: ${JSON.stringify(
+        (zodToJsonSchema(tool.schema) as JsonSchema7ObjectType).properties
```

Comment:
technically slightly different right now since the args schema in python is sub-selecting on `properties` but I think we're going to change that soon